File: ProjectEuler/Problems/Problem028.cs
namespace ProjectEuler.Problems
{
    public class Problem028 : Problem<long>
    {
        protected override long CalculateAnswer()
        {
            var n = 1001;
            var grid = new long[n, n];

            //populate grid with values
            var middle_index = (n - 1) / 2;
            var distance = 0;
            var distance_max = middle_index;
            var x = middle_index;
            var y = middle_index;
            var x_inc = 1;
            var y_inc = 0;
            var current_number = 1L;
            while (distance <= distance_max)
            {
                grid[x, y] = current_number;
                var distance_next = Distance(middle_index, x + x_inc, y + y_inc);
                if (distance_next > distance)
                {
                    //allow enlarging the distance by 1 only in the "right" direction
                    if (x_inc == 1 && distance_next == distance + 1)
                    {
                        ++distance;
                        x += x_inc;
                        y += y_inc;
                        ChangeDirection(ref x_inc, ref y_inc);
                    }
                    else
                    {
                        ChangeDirection(ref x_inc, ref y_inc);
                        x += x_inc;
                        y += y_inc;
                    }
                }
                else
                {
                    x += x_inc;
                    y += y_inc;
                }
                ++current_number;
            }

            var answer = 0L;
            //calculate diagonal sums
            for (var i = 0; i <= middle_index; ++i)
            {
                answer += grid[i, i];
                if (i != middle_index)
                {
                    answer += grid[i, n - i - 1];
                    answer += grid[n - i - 1, i];
                    answer += grid[n - i - 1, n - i - 1];
                }
            }

            return answer;
        }

        private int Distance(int middle_index, int x, int y)
        {
            var d1 = middle_index - x;
            if (d1 < 0) d1 = -d1;
            var d2 = middle_index - y;
            if (d2 < 0) d2 = -d2;
            if (d1 > d2) return d1;
            return d2;
        }

        private void ChangeDirection(ref int x_inc, ref int y_inc)
        {
            if (x_inc == 1 && y_inc == 0) //right
            {
                //down
                x_inc = 0;
                y_inc = 1;
            }
            else if (x_inc == 0 && y_inc == 1) //down
            {
                //left
                x_inc = -1;
                y_inc = 0;
            }
            else if (x_inc == -1 && y_inc == 0) //left
            {
                //up
                x_inc = 0;
                y_inc = -1;
            }
            else //up
            {
                //right
                x_inc = 1;
                y_inc = 0;
            }
        }
    }
}

File: ProjectEulerPython/problems/problem_028.py
from problems.problem import Problem

def distance(middle_index: int, x: int, y: int) -> int:
  dx = middle_index - x
  if dx < 0:
    dx = -dx
  dy = middle_index - y
  if dy < 0:
    dy = -dy
  if dx > dy:
    return dx
  return dy

def next_direction(dx: int, dy: int) -> (int, int):
  if dx == 1 and dy == 0: # right
    return (0, 1) # down
  if dx == 0 and dy == 1: # down
    return (-1, 0) # left
  if dx == -1 and dy == 0: # left
    return (0, -1) # up
  return (1, 0) # right

class Problem028(Problem):

  def calculate_answer(self) -> int:
    n = 1001
    grid = [] # 2 dimensional

    # fill with zeros
    for i in range(0, n):
      inner_grid = []
      for j in range(0, n):
        inner_grid.append(0)
      grid.append(inner_grid)

    middle_index = (n - 1) // 2
    d = 0
    d_max = middle_index
    x = middle_index
    y = middle_index
    dx = 1
    dy = 0
    current_number = 1

    # populate grid as described in the problem
    while d <= d_max:
      grid[x][y] = current_number
      d_next = distance(middle_index, x + dx, y + dy)
      if d_next > d:
        if dx == 1 and d_next == d + 1: # allow enlarging the distance by 1 only in the "right" direction
          d += 1
          x += dx
          y += dy
          dx, dy = next_direction(dx, dy)
        else: # greater than allowed distance, change direction
          dx, dy = next_direction(dx, dy)
          x += dx
          y += dy
      else: # continue to move in the same direction
        x += dx
        y += dy
      current_number += 1

    answer = 0
    # calculate diagonal sums
    for i in range(0, middle_index + 1):
      answer += grid[i][i]
      if i != middle_index:
        answer += grid[i][n - i - 1]
        answer += grid[n - i - 1][i]
        answer += grid[n - i - 1][n - i - 1]

    return answer